File: python/src/hypabolic_trajectory/io/file_stream.py
"""Path poll helpers → complete lines → core apply only (LS-09)."""

from __future__ import annotations

import os
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Iterator

from hypabolic_trajectory._enums import TrajectorySource
from hypabolic_trajectory.streaming.apply import TrajectoryStream
from hypabolic_trajectory.streaming.framing import split_complete_lines
from hypabolic_trajectory.streaming.types import StreamOptions, StreamUpdate

HOST_ROOT_REQUIRED = "root_required"
HOST_PATH_REQUIRED = "path_required"
HOST_PATH_OUTSIDE_ROOT = "path_outside_root"
HOST_IO_PERMISSION = "io_permission"
HOST_IO_NOT_FOUND = "io_not_found"
HOST_IO_ERROR = "io_error"

_MSG_ROOT_REQUIRED = "File stream root is required."
_MSG_PATH_REQUIRED = "File stream path is required."
_MSG_PATH_OUTSIDE_ROOT = "File stream path is outside the explicit root."
_MSG_IO_PERMISSION = "File stream could not read the path (permission denied)."
_MSG_IO_NOT_FOUND = "File stream path was not found."
_MSG_IO_ERROR = "File stream I/O failed."


class FileStreamHostError(Exception):
    """Host-side I/O or configuration error (not a stream diagnostic)."""

    def __init__(self, code: str, message: str, *, path: str | None = None) -> None:
        super().__init__(message)
        self.code = code
        self.message = message
        # Path is for the calling process only; never copy into StreamUpdate.
        self.path = path


@dataclass(frozen=True, slots=True, kw_only=True)
class FileStreamOptions:
    """Explicit-root file follow options."""

    root: str | Path
    path: str | Path
    source: TrajectorySource | str
    group_id: str | None = None
    stream: StreamOptions | None = None
    poll_interval: float = 0.05
    reconcile_every: int = 0  # 0 = disabled; N = full snapshot every N polls after first
    source_revision: str = "file-0"


class FileTrajectoryStream:
    """Poll a single JSONL path and apply complete segments to core streaming."""

    def __init__(
        self,
        *,
        root: Path,
        path: Path,
        stream: TrajectoryStream,
        poll_interval: float,
        reconcile_every: int,
        source_revision: str,
    ) -> None:
        self._root = root
        self._path = path
        self._stream = stream
        self._poll_interval = poll_interval
        self._reconcile_every = reconcile_every
        self._source_revision = source_revision
        self._file_offset = 0
        self._host_pending = b""
        self._first = True
        self._polls = 0
        self._closed = False
        self._identity: tuple[int, int, int] | None = None

    @classmethod
    def open(cls, options: FileStreamOptions) -> FileTrajectoryStream:
        root_raw = options.root
        path_raw = options.path
        if root_raw is None or (isinstance(root_raw, str) and not root_raw.strip()):
            raise FileStreamHostError(HOST_ROOT_REQUIRED, _MSG_ROOT_REQUIRED)
        if path_raw is None or (isinstance(path_raw, str) and not path_raw.strip()):
            raise FileStreamHostError(HOST_PATH_REQUIRED, _MSG_PATH_REQUIRED)

        root = Path(root_raw).expanduser().resolve()
        path = Path(path_raw).expanduser().resolve()
        if not _is_under_root(root, path):
            raise FileStreamHostError(
                HOST_PATH_OUTSIDE_ROOT,
                _MSG_PATH_OUTSIDE_ROOT,
                path=str(path),
            )

        stream_opts = options.stream
        if stream_opts is None:
            stream_opts = StreamOptions(source=options.source, group_id=options.group_id)
        elif options.group_id is not None and stream_opts.group_id is None:
            stream_opts = StreamOptions(
                source=stream_opts.source,
                group_id=options.group_id,
                delivery=stream_opts.delivery,
                include_provisional=stream_opts.include_provisional,
                require_complete_lines=stream_opts.require_complete_lines,
                finalize_on_close=stream_opts.finalize_on_close,
                reorder=stream_opts.reorder,
                reset_policy=stream_opts.reset_policy,
                max_pending_bytes=stream_opts.max_pending_bytes,
                max_line_bytes=stream_opts.max_line_bytes,
                normalize=stream_opts.normalize,
                ahp_protocol_version=stream_opts.ahp_protocol_version,
            )

        return cls(
            root=root,
            path=path,
            stream=TrajectoryStream.create(stream_opts),
            poll_interval=max(0.0, float(options.poll_interval)),
            reconcile_every=max(0, int(options.reconcile_every)),
            source_revision=options.source_revision,
        )

    @property
    def path(self) -> Path:
        return self._path

    @property
    def root(self) -> Path:
        return self._root

    @property
    def stream(self) -> TrajectoryStream:
        return self._stream

    @property
    def cursor(self):
        return self._stream.cursor

    def poll(self) -> StreamUpdate | None:
        """Read growth once; return a core StreamUpdate or None if unchanged."""
        if self._closed:
            return None
        size, identity = self._stat_identity()
        if size < self._file_offset:
            return self._snapshot_full(size, identity)
        if self._first:
            return self._snapshot_full(size, identity)
        if self._identity_changed(identity, size=size):
            # Inode/dev change or same-size rewrite (mtime) → full prefix.
            return self._snapshot_full(size, identity)
        if size > self._file_offset:
            return self._append_growth(size, identity)
        self._polls += 1
        if (
            self._reconcile_every > 0
            and self._polls % self._reconcile_every == 0
            and size >= 0
        ):
            return self._reconcile_snapshot(size, identity)
        self._identity = identity
        return None

    def follow(self, *, interval: float | None = None) -> Iterator[StreamUpdate]:
        """Yield non-empty updates until the process stops iterating.

        Caller owns lifetime (not a daemon). Use break/return to stop.
        """
        wait = self._poll_interval if interval is None else max(0.0, float(interval))
        while not self._closed:
            update = self.poll()
            if update is not None and update.kind != "unchanged":
                yield update
            if wait > 0:
                time.sleep(wait)

    def finish(self) -> StreamUpdate:
        """Finish the underlying core stream (does not close the file handle).

        Forwards any host-held incomplete line into core pending first so
        ``finish`` can commit a final unterminated line (core finish only sees
        core ``pending_bytes``).

        Host pending is retained until core ``apply_append`` succeeds. On
        non-success (error / reset-required / sequence-gap / resync-required),
        the pending buffer is kept, the failed update is returned, and
        ``finish`` is not applied.
        """
        if self._host_pending:
            update = self._stream.apply_append(
                self._host_pending, source_revision=self._source_revision
            )
            if update.kind not in ("updated", "unchanged"):
                return update
            self._host_pending = b""
        return self._stream.finish()

    def close(self) -> None:
        self._closed = True

    def __enter__(self) -> FileTrajectoryStream:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def _snapshot_full(
        self, size: int, identity: tuple[int, int, int]
    ) -> StreamUpdate:
        material = self._read_range(0, size)
        self._file_offset = size
        complete, pending = split_complete_lines(material)
        self._host_pending = pending
        self._first = False
        self._polls += 1
        self._identity = identity
        return self._stream.apply_snapshot(
            complete, source_revision=self._source_revision
        )

    def _reconcile_snapshot(
        self, size: int, identity: tuple[int, int, int]
    ) -> StreamUpdate | None:
        material = self._read_range(0, size)
        complete, pending = split_complete_lines(material)
        # Reconcile only when committed complete prefix is stable under host framing.
        self._host_pending = pending
        self._file_offset = size
        self._identity = identity
        update = self._stream.apply_snapshot(
            complete, source_revision=self._source_revision
        )
        if update.kind == "unchanged":
            return None
        return update

    def _append_growth(
        self, size: int, identity: tuple[int, int, int]
    ) -> StreamUpdate | None:
        chunk = self._read_range(self._file_offset, size)
        self._file_offset = size
        buf = self._host_pending + chunk
        complete, pending = split_complete_lines(buf)
        self._host_pending = pending
        self._polls += 1
        self._identity = identity
        if not complete:
            return None
        update = self._stream.apply_append(
            complete, source_revision=self._source_revision
        )
        if update.kind == "unchanged":
            return None
        return update

    def _identity_changed(self, identity: tuple[int, int, int], *, size: int) -> bool:
        if self._identity is None:
            return False
        # Inode/device change is authoritative (atomic replace). Same-size
        # in-place rewrite typically keeps the inode but updates mtime_ns.
        prev_dev, prev_ino, prev_mtime = self._identity
        dev, ino, mtime_ns = identity
        if (dev, ino) != (prev_dev, prev_ino):
            return True
        return size == self._file_offset and mtime_ns != prev_mtime

    def _stat_identity(self) -> tuple[int, tuple[int, int, int]]:
        try:
            st = os.stat(self._path)
            mtime_ns = getattr(st, "st_mtime_ns", int(st.st_mtime * 1_000_000_000))
            return int(st.st_size), (int(st.st_dev), int(st.st_ino), int(mtime_ns))
        except FileNotFoundError as exc:
            raise FileStreamHostError(
                HOST_IO_NOT_FOUND, _MSG_IO_NOT_FOUND, path=str(self._path)
            ) from exc
        except PermissionError as exc:
            raise FileStreamHostError(
                HOST_IO_PERMISSION, _MSG_IO_PERMISSION, path=str(self._path)
            ) from exc
        except OSError as exc:
            raise FileStreamHostError(
                HOST_IO_ERROR, _MSG_IO_ERROR, path=str(self._path)
            ) from exc

    def _read_range(self, start: int, end: int) -> bytes:
        if end <= start:
            return b""
        try:
            with open(self._path, "rb") as handle:
                handle.seek(start)
                return handle.read(end - start)
        except FileNotFoundError as exc:
            raise FileStreamHostError(
                HOST_IO_NOT_FOUND, _MSG_IO_NOT_FOUND, path=str(self._path)
            ) from exc
        except PermissionError as exc:
            raise FileStreamHostError(
                HOST_IO_PERMISSION, _MSG_IO_PERMISSION, path=str(self._path)
            ) from exc
        except OSError as exc:
            raise FileStreamHostError(
                HOST_IO_ERROR, _MSG_IO_ERROR, path=str(self._path)
            ) from exc


def _is_under_root(root: Path, path: Path) -> bool:
    try:
        path.relative_to(root)
        return True
    except ValueError:
        return False

File: typescript/packages/trajectory-node/src/file-stream.ts
/**
 * Optional file I/O for live session streaming (LS-09).
 * Poll helpers that only call core apply APIs. Explicit root required.
 */
import { open, stat } from "node:fs/promises";
import { isAbsolute, relative, resolve, sep } from "node:path";

import {
  applyAppend,
  applySnapshot,
  createStream,
  finishStream,
  splitCompleteLines as splitCompleteLinesCore,
  type StreamOptions,
  type StreamUpdate,
  type TrajectorySource,
} from "@hypabolic/trajectory";

export const HOST_ROOT_REQUIRED = "root_required" as const;
export const HOST_PATH_REQUIRED = "path_required" as const;
export const HOST_PATH_OUTSIDE_ROOT = "path_outside_root" as const;
export const HOST_IO_PERMISSION = "io_permission" as const;
export const HOST_IO_NOT_FOUND = "io_not_found" as const;
export const HOST_IO_ERROR = "io_error" as const;

const MSG_ROOT_REQUIRED = "File stream root is required.";
const MSG_PATH_REQUIRED = "File stream path is required.";
const MSG_PATH_OUTSIDE_ROOT = "File stream path is outside the explicit root.";
const MSG_IO_PERMISSION = "File stream could not read the path (permission denied).";
const MSG_IO_NOT_FOUND = "File stream path was not found.";
const MSG_IO_ERROR = "File stream I/O failed.";

export type FileStreamHostCode =
  | typeof HOST_ROOT_REQUIRED
  | typeof HOST_PATH_REQUIRED
  | typeof HOST_PATH_OUTSIDE_ROOT
  | typeof HOST_IO_PERMISSION
  | typeof HOST_IO_NOT_FOUND
  | typeof HOST_IO_ERROR;

export class FileStreamHostError extends Error {
  readonly code: FileStreamHostCode;
  /** For the calling process only; never copy into StreamUpdate diagnostics. */
  readonly path?: string;

  constructor(code: FileStreamHostCode, message: string, path?: string) {
    super(message);
    this.name = "FileStreamHostError";
    this.code = code;
    if (path !== undefined) this.path = path;
  }
}

export interface FileStreamOptions {
  readonly root: string;
  readonly path: string;
  readonly source: TrajectorySource;
  readonly groupId?: string;
  readonly stream?: StreamOptions;
  /** Poll interval seconds for followFile (default 0.05). */
  readonly pollInterval?: number;
  /** Full-prefix reconcile every N polls (0 = disabled). */
  readonly reconcileEvery?: number;
  readonly sourceRevision?: string;
}

/**
 * Host-edge complete-line split (same framing as core).
 * Property names `complete`/`pending` match the file-I/O package convention;
 * core exports `committed`/`pending`.
 */
export function splitCompleteLines(data: Uint8Array): {
  complete: Uint8Array;
  pending: Uint8Array;
} {
  const { committed, pending } = splitCompleteLinesCore(data);
  return { complete: committed, pending };
}

function isUnderRoot(root: string, path: string): boolean {
  const rel = relative(root, path);
  return rel === "" || (!rel.startsWith(`..${sep}`) && rel !== ".." && !isAbsolute(rel));
}

function mapIoError(error: unknown, path: string): FileStreamHostError {
  if (error instanceof FileStreamHostError) return error;
  if (error instanceof Error && "code" in error) {
    const code = (error as NodeJS.ErrnoException).code;
    if (code === "ENOENT") {
      return new FileStreamHostError(HOST_IO_NOT_FOUND, MSG_IO_NOT_FOUND, path);
    }
    if (code === "EACCES" || code === "EPERM") {
      return new FileStreamHostError(HOST_IO_PERMISSION, MSG_IO_PERMISSION, path);
    }
  }
  return new FileStreamHostError(HOST_IO_ERROR, MSG_IO_ERROR, path);
}

export class FileTrajectoryStream {
  readonly root: string;
  readonly path: string;
  #state: ReturnType<typeof createStream>;
  #fileOffset = 0;
  #hostPending: Uint8Array = new Uint8Array(0);
  #first = true;
  #polls = 0;
  #closed = false;
  #identity: { dev: number; ino: number; mtimeNs: bigint } | null = null;
  #reconcileEvery: number;
  #sourceRevision: string;
  #pollInterval: number;

  private constructor(
    root: string,
    path: string,
    state: ReturnType<typeof createStream>,
    pollInterval: number,
    reconcileEvery: number,
    sourceRevision: string,
  ) {
    this.root = root;
    this.path = path;
    this.#state = state;
    this.#pollInterval = pollInterval;
    this.#reconcileEvery = reconcileEvery;
    this.#sourceRevision = sourceRevision;
  }

  static open(options: FileStreamOptions): FileTrajectoryStream {
    if (!options.root || !options.root.trim()) {
      throw new FileStreamHostError(HOST_ROOT_REQUIRED, MSG_ROOT_REQUIRED);
    }
    if (!options.path || !options.path.trim()) {
      throw new FileStreamHostError(HOST_PATH_REQUIRED, MSG_PATH_REQUIRED);
    }
    const root = resolve(options.root);
    const path = resolve(options.path);
    if (!isUnderRoot(root, path)) {
      throw new FileStreamHostError(HOST_PATH_OUTSIDE_ROOT, MSG_PATH_OUTSIDE_ROOT, path);
    }
    let streamOpts: StreamOptions;
    if (options.stream !== undefined) {
      streamOpts =
        options.groupId !== undefined && options.stream.groupId === undefined
          ? { ...options.stream, groupId: options.groupId }
          : options.stream;
    } else if (options.groupId !== undefined) {
      streamOpts = { source: options.source, groupId: options.groupId };
    } else {
      streamOpts = { source: options.source };
    }
    return new FileTrajectoryStream(
      root,
      path,
      createStream(streamOpts),
      options.pollInterval ?? 0.05,
      options.reconcileEvery ?? 0,
      options.sourceRevision ?? "file-0",
    );
  }

  get cursor() {
    return this.#state.cursor;
  }

  get state() {
    return this.#state;
  }

  async poll(): Promise<StreamUpdate | null> {
    if (this.#closed) return null;
    const { size, identity } = await this.#statIdentity();
    if (size < this.#fileOffset) {
      return this.#snapshotFull(size, identity);
    }
    if (this.#first) {
      return this.#snapshotFull(size, identity);
    }
    if (this.#identityChanged(identity, size)) {
      return this.#snapshotFull(size, identity);
    }
    if (size > this.#fileOffset) {
      return this.#appendGrowth(size, identity);
    }
    this.#polls += 1;
    if (this.#reconcileEvery > 0 && this.#polls % this.#reconcileEvery === 0) {
      return this.#reconcileSnapshot(size, identity);
    }
    this.#identity = identity;
    return null;
  }

  /**
   * Yield non-empty updates until closed or AbortSignal aborts.
   * Abort/cancel stops the follow loop only; it does not call `finish()`.
   * Call `finish()` explicitly for end-of-stream (flush host pending + core finish).
   */
  async *follow(signal?: AbortSignal): AsyncGenerator<StreamUpdate, void, void> {
    const waitMs = Math.max(0, this.#pollInterval * 1000);
    while (!this.#closed && !(signal?.aborted ?? false)) {
      const update = await this.poll();
      if (update !== null && update.kind !== "unchanged") {
        yield update;
      }
      if (waitMs > 0) {
        await sleep(waitMs, signal);
      }
    }
  }

  /**
   * Finish the underlying core stream. Forwards host-held incomplete line into
   * core pending first so finish can commit a final unterminated line.
   * Host pending is retained until core apply succeeds; non-success returns
   * without calling finish. Distinct from AbortSignal/follow cancellation and
   * from `close()`.
   */
  finish(): StreamUpdate {
    if (this.#hostPending.length > 0) {
      const result = applyAppend(
        this.#state,
        this.#hostPending,
        undefined,
        this.#sourceRevision,
      );
      this.#state = result.state;
      if (result.update.kind !== "updated" && result.update.kind !== "unchanged") {
        return result.update;
      }
      this.#hostPending = new Uint8Array(0);
    }
    const finished = finishStream(this.#state);
    this.#state = finished.state;
    return finished.update;
  }

  close(): void {
    this.#closed = true;
  }

  async #snapshotFull(
    size: number,
    identity: { dev: number; ino: number; mtimeNs: bigint },
  ): Promise<StreamUpdate> {
    const material = await this.#readRange(0, size);
    this.#fileOffset = size;
    const { complete, pending } = splitCompleteLines(material);
    this.#hostPending = copyBytes(pending);
    this.#first = false;
    this.#polls += 1;
    this.#identity = identity;
    const result = applySnapshot(this.#state, complete, this.#sourceRevision);
    this.#state = result.state;
    return result.update;
  }

  async #reconcileSnapshot(
    size: number,
    identity: { dev: number; ino: number; mtimeNs: bigint },
  ): Promise<StreamUpdate | null> {
    const material = await this.#readRange(0, size);
    const { complete, pending } = splitCompleteLines(material);
    this.#hostPending = copyBytes(pending);
    this.#fileOffset = size;
    this.#identity = identity;
    const result = applySnapshot(this.#state, complete, this.#sourceRevision);
    this.#state = result.state;
    if (result.update.kind === "unchanged") return null;
    return result.update;
  }

  async #appendGrowth(
    size: number,
    identity: { dev: number; ino: number; mtimeNs: bigint },
  ): Promise<StreamUpdate | null> {
    const chunk = await this.#readRange(this.#fileOffset, size);
    this.#fileOffset = size;
    const buf = concatBytes(this.#hostPending, chunk);
    const { complete, pending } = splitCompleteLines(buf);
    this.#hostPending = copyBytes(pending);
    this.#polls += 1;
    this.#identity = identity;
    if (complete.length === 0) return null;
    const result = applyAppend(this.#state, complete, undefined, this.#sourceRevision);
    this.#state = result.state;
    if (result.update.kind === "unchanged") return null;
    return result.update;
  }

  #identityChanged(
    identity: { dev: number; ino: number; mtimeNs: bigint },
    size: number,
  ): boolean {
    if (this.#identity === null) return false;
    if (identity.dev !== this.#identity.dev || identity.ino !== this.#identity.ino) {
      return true;
    }
    return size === this.#fileOffset && identity.mtimeNs !== this.#identity.mtimeNs;
  }

  async #statIdentity(): Promise<{
    size: number;
    identity: { dev: number; ino: number; mtimeNs: bigint };
  }> {
    try {
      const info = await stat(this.path);
      const ns = (info as { mtimeNs?: bigint }).mtimeNs;
      const mtimeNs =
        typeof ns === "bigint" ? ns : BigInt(Math.round(info.mtimeMs * 1e6));
      return {
        size: info.size,
        identity: { dev: info.dev, ino: info.ino, mtimeNs },
      };
    } catch (error) {
      throw mapIoError(error, this.path);
    }
  }

  async #readRange(start: number, end: number): Promise<Uint8Array> {
    if (end <= start) return new Uint8Array(0);
    try {
      const handle = await open(this.path, "r");
      try {
        const length = end - start;
        const buffer = Buffer.alloc(length);
        const { bytesRead } = await handle.read(buffer, 0, length, start);
        return new Uint8Array(buffer.buffer, buffer.byteOffset, bytesRead);
      } finally {
        await handle.close();
      }
    } catch (error) {
      throw mapIoError(error, this.path);
    }
  }
}

export function openFileStream(options: FileStreamOptions): FileTrajectoryStream {
  return FileTrajectoryStream.open(options);
}

export async function* followFile(
  options: FileStreamOptions,
  signal?: AbortSignal,
): AsyncGenerator<StreamUpdate, void, void> {
  const stream = FileTrajectoryStream.open(options);
  try {
    yield* stream.follow(signal);
  } finally {
    stream.close();
  }
}

function concatBytes(a: Uint8Array, b: Uint8Array): Uint8Array {
  const out = new Uint8Array(a.length + b.length);
  out.set(a, 0);
  out.set(b, a.length);
  return out;
}

function copyBytes(data: Uint8Array): Uint8Array {
  const out = new Uint8Array(data.length);
  out.set(data);
  return out;
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolvePromise, reject) => {
    if (signal?.aborted) {
      reject(signal.reason ?? new Error("aborted"));
      return;
    }
    const timer = setTimeout(() => resolvePromise(), ms);
    signal?.addEventListener(
      "abort",
      () => {
        clearTimeout(timer);
        reject(signal.reason ?? new Error("aborted"));
      },
      { once: true },
    );
  });
}
